src/main/java/de/tum/in/www1/artemis/web/rest/GenericSubmissionResource.java

```diff
@@ -0,0 +1,166 @@
+ */
+public abstract class GenericSubmissionResource<T extends Submission, E extends Exercise> {
+
+    @Value("${jhipster.clientApp.name}")
+    protected String applicationName;
+
+    private static final String ENTITY_NAME = "genericSubmission";
+
+    protected final CourseService courseService;
+
+    protected final AuthorizationCheckService authCheckService;
+
+    protected final ExerciseService exerciseService;
+
+    protected final UserService userService;
+
+    protected final ParticipationService participationService;
+
+    public GenericSubmissionResource(CourseService courseService, AuthorizationCheckService authCheckService, UserService userService, ExerciseService exerciseService,
+            ParticipationService participationService) {
+        this.userService = userService;
+        this.exerciseService = exerciseService;
+        this.courseService = courseService;
+        this.authCheckService = authCheckService;
+        this.participationService = participationService;
+    }
+
+    /**
+     * Check if exercise is valid, has course and the user (student) can access it
+     * @param exercise that we want to check
+     * @return either null if exercise is valid or one of the error responses if it is not valid
+     */
+    ResponseEntity<T> checkExerciseValidityForStudent(E exercise) {
+        if (exercise == null) {
+            return ResponseEntity.badRequest()
+                    .headers(HeaderUtil.createFailureAlert(applicationName, true, "submission", "exerciseNotFound", "No exercise was found for the given ID.")).body(null);
+        }
+
+        // fetch course from database to make sure client didn't change groups
+        final Course course = courseService.findOne(exercise.getCourse().getId());
+        if (course == null) {
+            return ResponseEntity.badRequest()
+                    .headers(
+                            HeaderUtil.createFailureAlert(applicationName, true, ENTITY_NAME, "courseNotFound", "The course belonging to this file upload exercise does not exist"))
+                    .body(null);
+        }
+        if (!authCheckService.isAtLeastStudentInCourse(course, userService.getUserWithGroupsAndAuthorities())) {
+            return ResponseEntity.status(HttpStatus.FORBIDDEN).build();
+        }
+
+        return null;
+    }
+
+    /**
+     * Check if exercise is valid, the user (tutor) can access it and the user didn't reach the limit of submission locks
+     * @param exercise that we want to check
+     * @param exerciseType type of the exercise
+     * @param submissionService concrete submission service that is used to check lock limit
+     * @param <S> concrete submission repository class
+     * @return either null if exercise is valid or one of the error responses if it is not valid
+     */
+    <S extends GenericSubmissionRepository<T>> ResponseEntity<T> checkExerciseValidityForTutor(Exercise exercise, Class<E> exerciseType,
```

Comment:
I think, it should be possible to remove all generic dependencies like `<S extends GenericSubmissionRepository<T>>`. Generally, those always end up only being dependent on the Submission type itself. An example:
`public abstract class SubmissionService<T extends Submission, E extends GenericSubmissionRepository<T>>`
You can just write 
`public abstract class SubmissionService<T extends Submission>`
and have `GenericSubmissionRepository<T>` as the type of the injected bean instead of `E`. If you consistently do this, you want have to specify these complex generic type and will only end up needing one Generic (i.e. the Submission) most of the time

Reply:
I see that I used generic types too much in some cases. I removed unnecessary usage of generics and simplified SubmissionService to only require to specify submission type